core_java/grep/src/main/java/ca/jrvs/apps/grep/JavaGrepImp.java

```diff
@@ -0,0 +1,169 @@
+import org.slf4j.LoggerFactory;
+
+public class JavaGrepImp implements JavaGrep {
+
+    final Logger logger = LoggerFactory.getLogger(JavaGrep.class);
+
+    private String regex;
+    private String rootPath;
+    private String outFile;
+
+    @Override
+    public String getRegex() {
+        return regex;
+    }
+
+    @Override
+    public void setRegex(String regex) {
+        this.regex = regex;
+    }
+
+    @Override
+    public String getRootPath() {
+        return rootPath;
+    }
+
+    @Override
+    public void setRootPath(String rootPath) {
+        this.rootPath = rootPath;
+    }
+
+    @Override
+    public String getOutFile() {
+        return outFile;
+    }
+
+    @Override
+    public void setOutFile(String outFile) {
+        this.outFile = outFile;
+    }
+
+    /**
+     * Top level search
+     * @throws IOException
+     */
+    @Override
+    public void process() throws IOException {
+        List<String> matchedLines = new ArrayList<String>();
+
+        for (File file : listFiles(getRootPath())) {
+            for (String line : readLines(file)) {
+                if (containsPattern(line)) {
+                    matchedLines.add(line);
+                }
+            }
+        } try {
+            writeToFile(matchedLines);
+        } catch (IOException ex) {
+            logger.error("Unable to write to file.", ex);
+        }
+    }
+
+    /**
+     * Traverse a given directory and return all files
+     * @param rootDir input directory
+     * @return files under the rootDir
+     */
+    @Override
+    public List<File> listFiles(String rootDir) {
+        File[] tree = new File(rootDir).listFiles();
+        List<File> fileList = new ArrayList<File>();
+        if (tree != null) {
+            for (File file : tree) {
+                if (file.isDirectory()) {
+                    for (File dirFile : listFiles(file.getAbsolutePath())) {
+                        fileList.add(dirFile);
+                    }
+                }
+                else {
+                    fileList.add(file);
+                }
+            }
+        }
+
+        return fileList;
+    }
+
+    /**
+     * Read a file and return all the lines
+     *
+     * @param inputFile file to be read
+     * @return lines
+     * @throws IllegalArgumentException if a given input file is not a file
+     */
+    @Override
+    public List<String> readLines(File inputFile) {
+
+        if (!inputFile.isFile()) {
+            throw new IllegalArgumentException("Error! The input file: " + inputFile.getName() + " is not a file. Please enter a valid file.");
+        }
+
+        List<String> lines = new ArrayList<String>();
+        try (BufferedReader br = new BufferedReader(new FileReader(inputFile))) {
+            String line = "";
+            while ((line = br.readLine()) != null) {
+                lines.add(line);
+            }
+        } catch (IOException ex) {
+            logger.error(ex.getMessage(), ex);
```

Comment:
the ex.getMessage() is already part of the ex object (second argument). Therefore, please write some meaningful message instead. 
please fix this for all occurrences.